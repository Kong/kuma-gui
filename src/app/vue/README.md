---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

