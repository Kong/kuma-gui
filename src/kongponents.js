import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
