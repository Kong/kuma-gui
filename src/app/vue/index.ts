import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
