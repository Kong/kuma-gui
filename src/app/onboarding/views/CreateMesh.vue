<template>
  <RouteView
    v-slot="{ can }"
  >
    <RouteTitle
      :title="t('onboarding.routes.create-mesh.title')"
    />
    <AppView>
      <OnboardingPage>
        <template #header>
          <OnboardingHeading>
            <template #title>
              Create the mesh
            </template>
          </OnboardingHeading>
        </template>

        <template #content>
          <p class="mb-4 text-center">
            When you install, {{ t('common.product.name') }} creates a <i>default</i> mesh, but you can add as many meshes as you need.
          </p>

          <KTable
            class="table"
            :fetcher="() => tableData"
            :headers="TABLE_HEADERS"
            disable-pagination
          />

          <p class="mt-4 text-center">
            This mesh is empty. Next, you add services and their data plane proxies.
          </p>
        </template>

        <template #navigation>
          <OnboardingNavigation
            next-step="onboarding-add-services"
            :previous-step="can('use zones') ? 'onboarding-multi-zone' : 'onboarding-configuration-types'"
          />
        </template>
      </OnboardingPage>
    </AppView>
  </RouteView>
</template>

<script lang="ts" setup>
import { KTable } from '@kong/kongponents'
import { ref } from 'vue'

import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingPage from '../components/OnboardingPage.vue'
import AppView from '@/app/application/components/app-view/AppView.vue'
import RouteTitle from '@/app/application/components/route-view/RouteTitle.vue'
import RouteView from '@/app/application/components/route-view/RouteView.vue'
import { useI18n } from '@/utilities'

const TABLE_HEADERS = [
  { label: 'Name', key: 'name' },
  { label: 'Services', key: 'servicesAmount' },
  { label: 'DPPs', key: 'dppsAmount' },
]

const { t } = useI18n()

const tableData = ref<{ total: number, data: any [] }>({
  total: 1,
  data: [
    {
      name: 'default',
      servicesAmount: 0,
      dppsAmount: 0,
    },
  ],
})
</script>

<style lang="scss" scoped>
.table {
  width: 50%;
  margin: 0 auto;
}
</style>
