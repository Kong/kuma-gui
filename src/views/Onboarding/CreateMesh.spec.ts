import { createStore } from 'vuex'
import { render, screen } from '@testing-library/vue'
import { KButton, KTable } from '@kong/kongponents'

import { flushPromises } from '@vue/test-utils'
import CreateMesh from './CreateMesh.vue'

function renderComponent() {
  const store = createStore({
    modules: {
      config: {
        namespaced: true,
        state: {
          clientConfig: {
            mode: 'global',
            store: {
              type: 'memory',
            },
          },
        },
        getters: {
          getConfigurationType: (state) => state.clientConfig?.store?.type,
          getMulticlusterStatus: () => true,
        },
      },
    },
  })

  return render(CreateMesh, {
    global: {
      plugins: [store],
      components: {
        KButton,
        KTable,
      },
      stubs: {
        routerLink: {
          props: ['to'],
          template: '<a>{{ to.name }}</a>',
        },
      },
    },
  })
}

describe('CreateMesh.vue', () => {
  it('renders snapshot', async () => {
    const { container } = renderComponent()

    await flushPromises()

    expect(container).toMatchSnapshot()
  })

  it('renders multizone next step', () => {
    renderComponent()

    expect(screen.getByText('onboarding-multi-zone')).toBeInTheDocument()
  })
})
