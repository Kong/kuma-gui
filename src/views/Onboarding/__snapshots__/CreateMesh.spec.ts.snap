// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMesh.vue renders snapshot 1`] = `
<div>
  <div>
    <div
      class="onboarding-container"
    >
      <div
        class="onboarding-container__header"
      >
        
        <div
          class="relative"
        >
          <h1
            class="onboarding-title"
          >
            Create the mesh
          </h1>
          <!--v-if-->
        </div>
        
      </div>
      <div
        class="onboarding-container__content"
      >
        <div
          class="w-full"
        >
          
          <p
            class="text-center mb-4"
          >
             When you install, Kuma creates a 
            <i>
              default
            </i>
             mesh, but you can add as many meshes as you need. 
          </p>
          <div
            class="flex justify-center mt-10 mb-12 pb-12"
          >
            <div
              class="w-full sm:w-3/5 lg:w-2/5 p-4"
            >
              <div
                data-v-222fefa6=""
                is-small=""
              >
                <section
                  class="k-table-wrapper"
                  data-v-222fefa6=""
                >
                  <table
                    class="has-hover k-table"
                    data-tableid="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                    data-v-222fefa6=""
                  >
                    <thead
                      class=""
                      data-v-222fefa6=""
                    >
                      <tr
                        class=""
                        data-v-222fefa6=""
                      >
                        
                        <th
                          class=""
                          data-v-222fefa6=""
                        >
                          <span
                            class="d-flex align-items-center"
                            data-v-222fefa6=""
                          >
                            
                            <span
                              class=""
                              data-v-222fefa6=""
                            >
                              Name
                            </span>
                            
                            <!---->
                          </span>
                        </th>
                        <th
                          class=""
                          data-v-222fefa6=""
                        >
                          <span
                            class="d-flex align-items-center"
                            data-v-222fefa6=""
                          >
                            
                            <span
                              class=""
                              data-v-222fefa6=""
                            >
                              Services
                            </span>
                            
                            <!---->
                          </span>
                        </th>
                        <th
                          class=""
                          data-v-222fefa6=""
                        >
                          <span
                            class="d-flex align-items-center"
                            data-v-222fefa6=""
                          >
                            
                            <span
                              class=""
                              data-v-222fefa6=""
                            >
                              DPPs
                            </span>
                            
                            <!---->
                          </span>
                        </th>
                        
                      </tr>
                    </thead>
                    <tbody
                      data-v-222fefa6=""
                    >
                      
                      <tr
                        data-v-222fefa6=""
                      >
                        
                        <td
                          data-v-222fefa6=""
                        >
                          
                          default
                          
                        </td>
                        <td
                          data-v-222fefa6=""
                        >
                          
                          0
                          
                        </td>
                        <td
                          data-v-222fefa6=""
                        >
                          
                          0
                          
                        </td>
                        
                      </tr>
                      
                    </tbody>
                  </table>
                  <!---->
                </section>
              </div>
            </div>
          </div>
          <p
            class="text-center"
          >
             This mesh is empty. Next, you add services and their data plane proxies. 
          </p>
          
        </div>
      </div>
      
      <div
        class="mt-4 flex items-center flex-col sm:flex-row justify-between"
      >
        <a
          class="medium rounded primary k-button navigation-button navigation-button--back"
          data-v-6f773baa=""
          type="button"
        >
          onboarding-configuration-types
        </a>
        <div>
          <button
            class="small rounded btn-link k-button skip-button"
            data-v-6f773baa=""
            type="button"
          >
            
            <!---->
            
            
             Skip Setup 
            
            <!---->
          </button>
          <span
            class="inline-block"
          >
            <a
              class="medium rounded primary k-button navigation-button navigation-button--next"
              data-v-6f773baa=""
              type="button"
            >
              onboarding-add-services
            </a>
          </span>
        </div>
      </div>
      
    </div>
    <div
      class="background-image"
    />
  </div>
</div>
`;
